### PR TITLE
Landlock v1.1

### PR DESCRIPTION
**Pull request by regit**
Initial implementation of landlock support for Suricata.

- [x] I have read the contributing guide lines at https://redmine.openinfosecfoundation.org/projects/suricata/wiki/Contributing
- [x] I have signed the Open Information Security Foundation contribution agreement at https://suricata.io/about/contribution-agreement/
- [x] I have updated the user guide (in doc/userguide/) to reflect the changes made (if applicable)

Link to [redmine](https://redmine.openinfosecfoundation.org/projects/suricata/issues) ticket: https://redmine.openinfosecfoundation.org/issues/5479

Describe changes:
- Add landlock support
- Document it

### CODE REVIEW

**Comment by suricata-qa** [comment]
Information: QA ran without warnings. 

Pipeline 8474

File: suricata.yaml.in [hunk 1]
```diff
@@ -1091,6 +1091,18 @@ asn1-max-frames: 256
 #  user: suri
 #  group: suri
 
+# Use landlock security module under Linux
+landlock:
+  enabled: no
+  directories:
+    write:
+      - @e_logdir@
```
Comment (l0kod) [inline suricata.yaml.in:1099]: Couldn't the log and run directory be a static/define string?

File: src/util-landlock.c [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+{
+    prctl(PR_SET_NO_NEW_PRIVS, 1, 0, 0, 0);
+    if (landlock_restrict_self(ruleset_fd, 0))
+        SCLogError(SC_ERR_CONF_YAML_ERROR, "Can't self restrict: %s", strerror(errno));
+}
+
+static int LandlockSandboxingAddRule(int ruleset_fd, const char *directory, uint64_t permission)
+{
+    struct landlock_path_beneath_attr path_beneath = {
+        .allowed_access = permission,
```
Comment (l0kod) [inline src/util-landlock.c:84]: This should be masked with `LANDLOCK_ACCESS_FS_REFER` according to the Landlock ABI, as done with `ruleset_attr.handled_access_fs`. It might be easier to wrap ruleset_fd in a struct with the supported access rights.

**Review by l0kod — COMMENTED** [review]
Nice!

I guess you would like to squash non-doc related changes from the second commit to the first. ;)

File: src/util-landlock.c [hunk 3]
```diff
@@ -0,0 +1,189 @@
...
+    /* Read configuration variable and exit if no enforcement */
+    int conf_status;
+    ConfGetBool("landlock.enabled", &conf_status);
+    if (!conf_status) {
+        SCLogNotice("Landlock is not enabled in configuration");
+        return;
+    }
+    int ruleset_fd = LandlockCreateRuleset();
+
+    LandlockSandboxingWritePath(ruleset_fd, ConfigGetLogDirectory());
```
Comment (jasonish) [inline src/util-landlock.c:132]: If we are implicitly adding the log directory, do we need to include it in `landlock.directories.write`?

File: src/util-landlock.c [hunk 4]
```diff
@@ -0,0 +1,189 @@
...
+    LandlockSandboxingAddRule(ruleset_fd, ConfigGetDataDirectory(),
+            _LANDLOCK_SURI_ACCESS_FS_WRITE | _LANDLOCK_ACCESS_FS_READ);
```
Comment (jasonish) [inline src/util-landlock.c:134]: Currently this directory does not exist on a default install, can we check if it exists before adding the rule to prevent logging an error message?

**Review by jasonish — CHANGES_REQUESTED** [review]
Overall I like this idea.  Reminds me a bit of `pledge` on OpenBSD.

At some point I wonder if it makes sense to just do as much as possible automatically by default?  Outside of Lua scripts and plugins, we know what we need to read and write.

Is read of `/usr/` required for shared libraries and such?

File: suricata.yaml.in (hunk 1 above)
Comment (victorjulien) [reply suricata.yaml.in:1099]: the are set by `--sysconfdir` and `--localstatedir`, e.g. `./configure --prefix=/usr --sysconfdir=/etc --localstatedir=/var`

**Comment by victorjulien** [comment]
Thanks for jumping for review @l0kod !

**Comment by regit** [comment]
> Nice!
> 
> I guess you would like to squash non-doc related changes from the second commit to the first. ;)

Damn, the clang rewrite script did screw me up once more.

File: src/util-landlock.h [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+{
+    return syscall(__NR_landlock_restrict_self, ruleset_fd, flags);
+}
+#endif
+
+#ifndef LANDLOCK_ACCESS_FS_REFER
+#define LANDLOCK_ACCESS_FS_REFER (1ULL << 13)
+#endif
+
+#endif /* HAVE_LINUX_LANDLOCK_H */
```
[the remark this answers is not in the document]
Comment (regit) [reply src/util-landlock.h:62]: Yes, that should work.

File: suricata.yaml.in [hunk 6]
```diff
@@ -1091,6 +1091,18 @@ asn1-max-frames: 256
...
+      - @e_rundir@
+    read:
+      - /usr/
```
[the remark this answers is not in the document]
Comment (regit) [reply suricata.yaml.in:1102]: I did put it there as a feature like filemagic is reading its config file in /etc and load things from /usr/.

File: src/util-landlock.c [hunk 7]
```diff
@@ -0,0 +1,189 @@
...
+void LandlockSandboxing(SCInstance *suri)
+{
+    /* Read configuration variable and exit if no enforcement */
+    int conf_status;
+    ConfGetBool("landlock.enabled", &conf_status);
+    if (!conf_status) {
+        SCLogNotice("Landlock is not enabled in configuration");
+        return;
+    }
+    int ruleset_fd = LandlockCreateRuleset();
```
[the remark this answers is not in the document]
Comment (regit) [reply src/util-landlock.c:130]: Indeed.

File: suricata.yaml.in (hunk 1 above)
Comment (regit) [reply suricata.yaml.in:1099]: These values it can be overwritten from configuration file so I prefer to have them here. For the log directory @jasonish remark on -l usage makes sense and it could be better to avoid to add it in the configuration.

File: src/util-landlock.c [hunk 8]
```diff
@@ -0,0 +1,189 @@
...
+    }
+    if (!suri->sig_file_exclusive) {
+        const char *rule_path;
+        ConfGet("default-rule-path", &rule_path);
+        if (rule_path) {
+            LandlockSandboxingReadPath(ruleset_fd, rule_path);
+        }
+    }
+
+    ConfNode *read_dirs = ConfGetNode("landlock.directories.read");
```
[the remark this answers is not in the document]
Comment (regit) [reply src/util-landlock.c:160]: It is possible to specify full path in signature for example for dataset, lua script. As a user can't really rewrite the signatures, this feature is needed.

**Comment by victorjulien** [comment]
> > Nice!
> > I guess you would like to squash non-doc related changes from the second commit to the first. ;)
> 
> Damn, the clang rewrite script did screw me up once more.

`git rebase master -x 'git clang-format HEAD'` works well for me

File: src/util-landlock.c (hunk 3 above)
Comment (regit) [reply src/util-landlock.c:132]: Yes, this is not needed. Good catch!

File: src/util-landlock.c (hunk 4 above)
Comment (regit) [reply src/util-landlock.c:134]: Good idea.

**Comment by regit** [comment]
> Overall I like this idea. Reminds me a bit of `pledge` on OpenBSD.
> 
> At some point I wonder if it makes sense to just do as much as possible automatically by default? Outside of Lua scripts and plugins, we know what we need to read and write.
> 
> Is read of `/usr/` required for shared libraries and such?

Yes, it is needed for libmagic at least who wants `/etc` and `/usr`. Maybe this is the only one.

File: src/util-landlock.c (hunk 8 above)
[the remark this answers is not in the document]
Comment (victorjulien) [reply src/util-landlock.c:160]: I guess we can talk about the thread model. If we trust the config (+rules etc) at startup, we might be able collect all the paths we open and then lock things afterwards? Then we'd protect against runtime exploitation of some kind.

If we want to protect the config/rule loading stage as well we couldn't use this.

File: suricata.yaml.in (hunk 6 above)
[the remark this answers is not in the document]
Comment (victorjulien) [reply suricata.yaml.in:1102]: Is some thing like `/usr/*/magic.mgc` possible here?

[the remark this answers is not in the document]
Comment (regit) [reply suricata.yaml.in:1102]: I've got a nightmare of link on debian to access to this file. I don't think we can do something else than directory with landlock and certainly not a blob.

File: suricata.yaml.in (hunk 1 above)
Comment (regit) [reply suricata.yaml.in:1099]: I got in down to 
```yaml
  directories:
    #write:
    #  - @e_rundir@
    read:
      - /usr/
      - /etc/
      - @e_sysconfdir@
```

File: src/util-landlock.c (hunk 2 above)
Comment (regit) [reply src/util-landlock.c:84]: OK got it now!

File: src/util-landlock.c (hunk 8 above)
[the remark this answers is not in the document]
Comment (l0kod) [reply src/util-landlock.c:160]: Right, and we could do both: a first layer of generic rules to forbid everything that a lua script should never be allowed to do (e.g., create block/char devices, read sensitive files, write new rules…), and a second (standalone) layer taking into account the explicit scripts/rules requirements. Being able to not ask users to explicitly configure sandboxing should be preferred to get a wide adoption.

File: suricata.yaml.in (hunk 6 above)
[the remark this answers is not in the document]
Comment (l0kod) [reply suricata.yaml.in:1102]: All glob/regex resolution is the job of user space, and it is then freeze once loaded in the kernel.

File: suricata.yaml.in (hunk 1 above)
Comment (l0kod) [reply suricata.yaml.in:1099]: > the are set by `--sysconfdir` and `--localstatedir`, e.g. `./configure --prefix=/usr --sysconfdir=/etc --localstatedir=/var`

By static/define I meant C `static` or `#define` strings that could be populated by the `./configure` command at build time.

Comment (l0kod) [reply suricata.yaml.in:1099]: > I got in down to in the MR to come
> 
> ```yaml
>   directories:
>     #write:
>     #  - @e_rundir@
>     read:
>       - /usr/
>       - /etc/
>       - @e_sysconfdir@
> ```

Great to not have to worry about write permissions!

`/usr` and `/etc` could maybe be inferred from a build configuration, maybe a new one with these paths as default values? I think it would be OK if users only had to specify such paths when they have a non-standard installation or distro.

**Comment by l0kod** [comment]
About `LANDLOCK_ACCESS_FS_REFER`, you need to make sure it is not required by Suricata to (hard)link or rename files from/to different hierarchies. If it is legitimate for Lua scripts to do so, then we could defer the choice to enable Landlock with ABI 1 to users or infer from the Lua scripts. If it may be legitimate to do so for a default Suricata installation, you should abort sandboxing if the Landlock ABI is 1 (which would deny such actions except if the source/destination are in the same directory).

I guess link and rename actions are not common for most applications, but we should keep this in mind.

 `LANDLOCK_ACCESS_FS_REFER` was a bit complex to implement securely, then the incremental development, which is the reason why renaming/linking was partially restricted at first. This access right (and the lifted limitation) is supported since Linux 5.19.

TL; DR: If Suricata might legitimately rename or link files to different directories, you should consider only using Landlock with ABI >= 2.

See https://www.kernel.org/doc/html/latest/userspace-api/landlock.html#file-renaming-and-linking-abi-1

File: suricata.yaml.in (hunk 6 above)
[the remark this answers is not in the document]
Comment (victorjulien) [reply suricata.yaml.in:1102]: Gotcha. Trick with libmagic is that to maximize portability we by default don't specify a path to the magic file. The library will then load from a default value that differs between OS'/distros. I haven't seen a way to retrieve the path that it uses.

**Comment by regit** [comment]
> About `LANDLOCK_ACCESS_FS_REFER`, you need to make sure it is not required by Suricata to (hard)link or rename files from/to different hierarchies. If it is legitimate for Lua scripts to do so, then we could defer the choice to enable Landlock with ABI 1 to users or infer from the Lua scripts. If it may be legitimate to do so for a default Suricata installation, you should abort sandboxing if the Landlock ABI is 1 (which would deny such actions except if the source/destination are in the same directory).
> 
> I guess link and rename actions are not common for most applications, but we should keep this in mind.
> 
> `LANDLOCK_ACCESS_FS_REFER` was a bit complex to implement securely, then the incremental development, which is the reason why renaming/linking was partially restricted at first. This access right (and the lifted limitation) is supported since Linux 5.19.
> 
> TL; DR: If Suricata might legitimately rename or link files to different directories, you should consider only using Landlock with ABI >= 2.
> 
> See https://www.kernel.org/doc/html/latest/userspace-api/landlock.html#file-renaming-and-linking-abi-1

H

**Comment by regit** [comment]
> About `LANDLOCK_ACCESS_FS_REFER`, you need to make sure it is not required by Suricata to (hard)link or rename files from/to different hierarchies. If it is legitimate for Lua scripts to do so, then we could defer the choice to enable Landlock with ABI 1 to users or infer from the Lua scripts. If it may be legitimate to do so for a default Suricata installation, you should abort sandboxing if the Landlock ABI is 1 (which would deny such actions except if the source/destination are in the same directory).
> 
> I guess link and rename actions are not common for most applications, but we should keep this in mind.
> 
> `LANDLOCK_ACCESS_FS_REFER` was a bit complex to implement securely, then the incremental development, which is the reason why renaming/linking was partially restricted at first. This access right (and the lifted limitation) is supported since Linux 5.19.
> 
> TL; DR: If Suricata might legitimately rename or link files to different directories, you should consider only using Landlock with ABI >= 2.
> 
> See https://www.kernel.org/doc/html/latest/userspace-api/landlock.html#file-renaming-and-linking-abi-1

OK, this could be really an issue with files extraction, I need to test that.

**Comment by l0kod** [comment]
> OK, this could be really an issue with files extraction, I need to test that.

You can test that behavior by just removing `LANDLOCK_ACCESS_FS_REFER` from `handled_access_fs` (and then `allowed_access`).

**Comment by regit** [comment]
> > OK, this could be really an issue with files extraction, I need to test that.
> 
> You can test that behavior by just removing `LANDLOCK_ACCESS_FS_REFER` from `handled_access_fs` (and then `allowed_access`).

OK, I run a 5.18 and I got:
```
[68938] 4/8/2022 -- 16:43:59 - (output-filestore.c:144) <Warning> (OutputFilestoreFinalizeFiles) -- [ERRCODE: SC_WARN_RENAMING_FILE(303)] - Failed to rename /tmp/ddd//filestore/tmp/file.1 to /tmp/ddd//filestore/01/010f5a2db9b9ce1e8a9e0ae6837f8fb208f825423ab35aac09c26e3d2b3d4aef: Invalid cross-device link
[68942] 4/8/2022 -- 16:43:59 - (output-filestore.c:144) <Warning> (OutputFilestoreFinalizeFiles) -- [ERRCODE: SC_WARN_RENAMING_FILE(303)] - Failed to rename /tmp/ddd//filestore/tmp/file.2 to /tmp/ddd//filestore/ef/ef7239c674fd380fc2d8713d0fd82be91e543b7562ccfdffce3357819be1fd2c: Invalid cross-device link
```

So it makes sense for now to disable landlock if abi is 1 and filestore is enabled.

**Comment by l0kod** [comment]
> So it makes sense for now to disable landlock if abi is 1 and filestore is enabled.

Usually, tools copy files instead of renaming or linking them (e.g. mv), but in the case of filestore I agree with these conditions when to not use Landlock.

**Comment by l0kod** [comment]
> Is this related? https://redmine.openinfosecfoundation.org/issues/3003
> 
> Wonder if it could make sense to optionally support this using a copy logic.This would be at the cost of perf, but the trade off might be worth it to some?

Indeed, that might be interesting. Renaming and linking will always be faster than a copy, but the cost could be small using in-kernel copy such as sendfile or splice, especially with tmpfs.

**Comment by regit** [comment]
Replaced by #7696
